File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> list = new ArrayList<>();
        int startRow = 0,startCol = 0, endRow = matrix.length - 1, endCol = matrix[0].length - 1;
        while(startRow <= endRow && startCol <= endCol){
            // traverse top row
            for(int i=startCol;i<=endCol;i++){
                list.add(matrix[startRow][i]);
            }
            startRow++;
            // traversing right col
            for(int i=startRow;i<=endRow;i++){
                list.add(matrix[i][endCol]);
            }
            endCol--;
            // traversing buttom row
            if(startRow <= endRow) {
            for(int i=endCol;i>=startCol;i--){
                list.add(matrix[endRow][i]);
            }
            endRow--;
            }
            // traversing left col
            if(startCol <= endCol) {
            for(int i=endRow;i>=startRow;i--){
                list.add(matrix[i][startCol]);
            }
            startCol++;
          }
        }
        return list;
    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> ans;
        int rs = 0, re = matrix.size() - 1;
        int cs = 0, ce = matrix[0].size() - 1;
        while(rs <= re && cs <= ce) {
            // traverse the top row
            for(int j=cs;j<=ce;j++) {
                ans.push_back(matrix[rs][j]);
            }
            rs++;
            // traverse the end column
            for(int i=rs;i<=re;i++) {
                ans.push_back(matrix[i][ce]);
            }
            ce--;
            // traverse the buttom row
            if(rs <= re) {
                for(int j=ce;j>=cs;j--){
                    ans.push_back(matrix[re][j]);
                }
                re--;
            }
            // traverse the left row
            if(cs <= ce) {
                for(int i=re;i>=rs;i--){
                    ans.push_back(matrix[i][cs]);
                }
                cs++;
            }
        }
        return ans;
    }
};